chain/traversal.go

```diff
@@ -0,0 +1,100 @@
+type BlockProvider interface {
+	GetBlock(ctx context.Context, cid cid.Cid) (*types.Block, error)
+}
+
+// GetParentTipSet returns the parent tipset of a tipset.
+// The result is empty if the tipset has no parents (including if it is empty itself)
+func GetParentTipSet(ctx context.Context, store BlockProvider, ts types.TipSet) (types.TipSet, error) {
+	newTipSet := types.TipSet{}
+	parents, err := ts.Parents()
+	if err != nil {
+		return nil, err
+	}
+	for it := parents.Iter(); !it.Complete() && ctx.Err() == nil; it.Next() {
+		newBlk, err := store.GetBlock(ctx, it.Value())
+		if err != nil {
+			return nil, err
+		}
+		if err := newTipSet.AddBlock(newBlk); err != nil {
+			return nil, err
+		}
+	}
+	return newTipSet, nil
+}
+
+// IterAncestors returns an iterator over tipset ancestors, yielding first the start tipset and
+// then its parent tipsets until (and including) the genesis tipset.
+func IterAncestors(ctx context.Context, store BlockProvider, start types.TipSet) *TipsetIterator {
+	return &TipsetIterator{ctx, store, start, nil}
+}
+
+// TipsetIterator is an iterator over tipsets.
+type TipsetIterator struct {
+	ctx   context.Context
+	store BlockProvider
+	value types.TipSet
+	err   error
+}
+
+// Value returns the iterator's current value, if not Complete().
+func (it *TipsetIterator) Value() types.TipSet {
+	return it.value
+}
+
+// Complete tests whether the iterator is exhausted.
+func (it *TipsetIterator) Complete() bool {
+	return len(it.value) == 0
+}
+
+// Next advances the iterator to the next value.
+func (it *TipsetIterator) Next() error {
+	it.value, it.err = GetParentTipSet(it.ctx, it.store, it.value)
+	return it.err
+}
+
+// IterAncestorBlocks returns an iterator over the blocks in a tipset and its ancestors until (and including)
+// the genesis.
+// Blocks from each tipset are yielded in an unspecified but stable order.
+// NOTE: it would be great to align this order with the canonical ordering and define it as such.
+func IterAncestorBlocks(ctx context.Context, store BlockProvider, start types.TipSet) *BlockIterator {
+	tsIter := start.ToSortedCidSet().Iter()
+	return &BlockIterator{TipsetIterator{ctx, store, start, nil}, tsIter}
+}
+
+// BlockIterator is an iterator over blocks in tipsets.
+type BlockIterator struct {
+	TipsetIterator
+	cidIterator types.SortedCidSetIterator
+}
+
+// Value returns the iterator's current value, if not Complete().
+func (it *BlockIterator) Value() *types.Block {
+	return it.value[it.cidIterator.Value()]
+}
+
+// Complete tests whether the iterator is exhausted.
+func (it *BlockIterator) Complete() bool {
+	return it.TipsetIterator.Complete() && it.cidIterator.Complete()
+}
+
+// Next advances the iterator to the next value.
+func (it *BlockIterator) Next() error {
+	it.cidIterator.Next()
+	if it.cidIterator.Complete() {
+		err := it.TipsetIterator.Next()
+		if err == nil {
+			it.cidIterator = it.TipsetIterator.Value().ToSortedCidSet().Iter()
```

Comment:
It is unclear why blocks are not returned in the canonical order if we are iterating over the tipset in sorted cid order.

Reply:
I couldn't find a bit of code that clearly identified this as the canonical order. Is this the canonical order?

Reply:
This is a good point.  In retrospect we aren't being good about enforcing one canonical order.  Tipsets are printed and serialized to slices via `ToSlice` in this order but more fundamentally their messages are serialized in the order of increasing ticket bytes.  I think it probably makes sense to enforce the ticket ordering everywhere.

Reply:
Agreed. I'll leave this and the comment about it here for now.